sklearn/linear_model/_ridge.py

```diff
@@ -2123,6 +2120,50 @@ def fit(self, X, y, sample_weight=None):
 
         return self
 
+    def _get_scorer(self):
+        return check_scoring(self, scoring=self.scoring, allow_none=True)
+
+    def _score_without_scorer(self, squared_errors):
```

Comment:
Added.

Reply:
I am wondering why don't we pass `score_params` to it?

Reply:
So for this case, we are in the case that we can only compute the MSE so there is no `score_params` that can be passed because we are in a single case. However, I'm concerned with `sample_weight` that is the only params that we could take into account for the GCV scheme:

https://github.com/scikit-learn/scikit-learn/blob/1d22a48aad533354ba8de25e2ad69cdfa11f11e3/sklearn/linear_model/_ridge.py#L2151-L2156

So I see that we use `sample_weight` to resolve the problem. However, the squared errors are not weighting with `sample_weight`. In the case, that we have a solver, I'm under the impression that we would reapply the weights again to the metric. Do I miss something?

Reply:
I think this was the default case that was present when we don't have any scorer. Here we calculate the squared errors simply using a line of code as is shown above. Do we require any params to correctly calculate the alpha_score using the squared errors?

Reply:
Do you mean we use `score_params` in the case when `self.scoring is not None`?

Reply:
> Do you mean we use score_params in the case when self.scoring is not None?
Yes, this is what I meant, sorry.

Reply:
I think it seems correct to use `sample_weight` in the other case too then, if it is present. The old behaviour should only be preserved if `sample_weight` is not passed.

Reply:
The following behaviour tells me that we should be OK when metadata is on.

```python
import numpy as np
import sklearn
from sklearn.datasets import make_regression
from sklearn.linear_model import RidgeCV
from sklearn.metrics import get_scorer

sklearn.set_config(enable_metadata_routing=True)

X, y = make_regression(random_state=0)
sample_weight = np.ones_like(y)
sample_weight[::2] = 0.5

ridge_no_scoring = RidgeCV(scoring=None).set_score_request(
    sample_weight=True
)
ridge_no_scoring.fit(
    X, y, sample_weight=sample_weight
)
scorer = get_scorer("neg_mean_squared_error")
ridge_with_scorer = RidgeCV(scoring=scorer).set_score_request(
    sample_weight=True
)
ridge_with_scorer.fit(
    X, y, sample_weight=sample_weight
)

print(
    f"Ridge with no scoring: {ridge_no_scoring.best_score_}\n"
    f"Ridge with a scorer: {ridge_with_scorer.best_score_}"
)
```

```
Ridge with no scoring: -419.01977944191776
Ridge with a scorer: -419.01977944191776
```

Reply:
Otherwise this is not the case:

```python
import numpy as np
import sklearn
from sklearn.datasets import make_regression
from sklearn.linear_model import RidgeCV
from sklearn.metrics import get_scorer

sklearn.set_config(enable_metadata_routing=False)

X, y = make_regression(random_state=0)
sample_weight = np.ones_like(y)
sample_weight[::2] = 0.5

ridge_no_scoring = RidgeCV(scoring=None)
ridge_no_scoring.fit(
    X, y, sample_weight=sample_weight
)
scorer = get_scorer("neg_mean_squared_error")
ridge_with_scorer = RidgeCV(scoring=scorer)
ridge_with_scorer.fit(
    X, y, sample_weight=sample_weight
)

print(
    f"Ridge with no scoring: {ridge_no_scoring.best_score_}\n"
    f"Ridge with a scorer: {ridge_with_scorer.best_score_}"
)
```

```
Ridge with no scoring: -419.01977944191776
Ridge with a scorer: -466.56603698219783
```

Reply:
So it means that in the past, the scorer was not taking into account the sample-weight while the internal GCV via the `c / gram_inverse_diag` is already taking the sample-weight into account.

Reply:
Actually we should make sure to have a test as above ;)

Reply:
This happens when the user hasn't passed any scorer. And that means we don't really have a scorer to route anything to.

However, there are a few issues here:

Here we pass `self.scoring` instead of `self._get_scorer()`:

https://github.com/scikit-learn/scikit-learn/blob/1d22a48aad533354ba8de25e2ad69cdfa11f11e3/sklearn/linear_model/_ridge.py#L2404-L2412

That means this probably doesn't do the right thing:

```py
est = RidgeCV().set_score_request(sample_weight=True).set_fit_request(sample_weight=True)
pipe = make_pipeline(est)
pipe.fit(X, y, sample_weight=sample_weight)
```

And if we pass `self._get_scorer()` instead of `self.scoring` to `_RidgeGCV`, then that `if scorer is None` will never happen, so it would clear things up quite a bit.

Note that this is visible now since @OmarManzoor nicely put all scoring code into methods, and now it's visible that we're not passing the right things around.

So I'd say this needs a fix.

Does this clear things up maybe? Happy to have a call if it doesn't.

Reply:
Another thing to make sure we check is, whether `sample_weight` is forwarded to `self.score` in all cases.

Reply:
> So it means that in the past, the scorer was not taking into account the sample-weight while the internal GCV via the c / gram_inverse_diag is already taking the sample-weight into account.

I just read this. Interesting! So my comment about passing `sample_weight` is kinda void here.

Reply:
I did not set the `sample_weight` on the score above so:

```python
import numpy as np
import sklearn
from sklearn.datasets import make_regression
from sklearn.linear_model import RidgeCV
from sklearn.metrics import get_scorer

sklearn.set_config(enable_metadata_routing=True)

X, y = make_regression(random_state=0)
sample_weight = np.ones_like(y)
sample_weight[::2] = 0.5

ridge_no_scoring = RidgeCV(scoring=None).set_score_request(
    sample_weight=True
)
ridge_no_scoring.fit(
    X, y, sample_weight=sample_weight
)
scorer = get_scorer("neg_mean_squared_error").set_score_request(
    sample_weight=True
)
ridge_with_scorer = RidgeCV(scoring=scorer).set_score_request(
    sample_weight=True
)
ridge_with_scorer.fit(
    X, y, sample_weight=sample_weight
)

print(
    f"Ridge with no scoring: {ridge_no_scoring.best_score_}\n"
    f"Ridge with a scorer: {ridge_with_scorer.best_score_}"
)
```

```
Ridge with no scoring: -419.01977944191776
Ridge with a scorer: -466.56603698219783
```

And I'm lost :)

Reply:
> And if we pass self._get_scorer() instead of self.scoring to _RidgeGCV, then that if scorer is None will never happen, so it would clear things up quite a bit.

This is exactly what you pushed last month in the other PR, see [commit](https://github.com/scikit-learn/scikit-learn/pull/29634/commits/3956f2cdacccfbd9bcabd3a9ff43ca935721ccdf). So, it's already taken care of, @adrinjalali

Reply:
Oh, then could you check Guillaume's test on your branch and see what it gives?

Reply:
Also, should we check for equivalence-ish between `RidgeCV` and `GridSearchCV(Ridge())`?

Reply:
On my branch, @glemaitres last test with `set_score_request` on the scorer there are same results as on main:

```
Ridge with no scoring: -419.0197794418993
Ridge with a scorer: -466.56603698217515
```
for both tests (with metadata routing and without).

Reply:
Ok, then I need to have a look.

Reply:
I'll craft a better dataset to know what should be the MSE: a perfect regression with a single outlier that should be removed with sample weight. So the true score should be zero.

Reply:
So here would be such an example:

```
import numpy as np
import sklearn
from sklearn.linear_model import RidgeCV
from sklearn.metrics import get_scorer
from sklearn.pipeline import make_pipeline

sklearn.set_config(enable_metadata_routing=True)

rng = np.random.default_rng(0)
coef = rng.uniform(size=(10,))
X = rng.normal(size=(1_000, 10))
y = X @ coef
y[-1] = 1_000  # add outlier
sample_weight = np.ones_like(y)
sample_weight[-1] = 0.0

ridge_no_scoring = make_pipeline(
    RidgeCV(scoring=None).set_score_request(
        sample_weight=True
    ).set_fit_request(sample_weight=True)
)
ridge_no_scoring.fit(
    X, y, sample_weight=sample_weight
)
scorer = get_scorer("neg_mean_squared_error").set_score_request(
    sample_weight=True
)
ridge_with_scorer = make_pipeline(
    RidgeCV(scoring=scorer).set_score_request(
        sample_weight=True
    ).set_fit_request(sample_weight=True)
)
ridge_with_scorer.fit(
    X, y, sample_weight=sample_weight
)

print(
    f"Ridge with no scoring: {ridge_no_scoring[-1].best_score_}\n"
    f"Ridge with a scorer: {ridge_with_scorer[-1].best_score_}"
)
np.testing.assert_allclose(
    ridge_no_scoring[-1].coef_, ridge_with_scorer[-1].coef_
)
```

```
Ridge with no scoring: -4.0321032019694904e-08
Ridge with a scorer: -4.036139341310801e-08
```

I would have expected a much bigger difference if we would have not take into account the weight.

Reply:
So looking closer, the following line is used to generate the predictions (as per written in http://cbcl.mit.edu/publications/ps/MIT-CSAIL-TR-2007-025.pdf)

```python
predictions = y - (c / G_inverse_diag)
```
However, if we use the prediction function from the `RidgeCV`, we would have something equivalent to:

```python
X @ safe_sparse_dot(c.T, X)
```

And the results are slightly different. I have to check more in details what is the reason.

Reply:
So forget what ever I said before and @StefanieSenger was right from the start, this is just an issue of the `sample_weight`. So if we want to have the same call than with a scorer, we would need the following function:

```python
    def _score_without_scorer(self, squared_errors, score_params):
        """Performs scoring using squared errors when the scorer is None."""
        axis = 0 if self.alpha_per_target else None
        weights = score_params.get("sample_weight", None)
        return -_average(squared_errors, axis=axis, weights=weights)
```

Reply:
I am confused about the premises to the discussion, which boils down to this question if `sample_weight` should be used consistently with metadata routing enabled only or also with it disabled.

When we enable metadata routing, do we want to mimic the old behaviour with it (meaning the different scores should re-appear) or would we route `sample_weight` even if it's not routed otherwise?

It seems that the desired result was to have the same scores if metadata routing is enabled, and accepting that it's not the same without the routing enabled, but then we were talking about fixing the inconsistency of the original code?

Reply:
So we got a small sync talk with @StefanieSenger but here is a summary of whatever happen in this thread :). Posting it here to make sure that the thread become clearer.

Yep the discussion went sideways. Sorry about that. I used this thread as a debugging thread :).

So to summarize, I'm not to worry about the consistency between disabling/enabling metadata routing because we might not reach the proper behaviour without metadata routing and meta-estimator. So here, what is really important is to have the right behaviour for when enabling metadata-routing. It means:

- when `scoring is None`, we need to make sure that `sample_weight` is taken into account. Basically, this is not configurable by the user since it does not pass a scorer.
- so the behaviour of `scoring is None` is equivalent than passing a scorer using the mean squared error and setting the score parameter to accept sample weight. We should be able to test this equivalence.

For me this is the behaviour that we should make sure is right. So it results in adding `sample_weight` or `score_params` to the `_score_without_scorer`.